rllib/utils/replay_buffers/episode_replay_buffer.py

```diff
@@ -0,0 +1,508 @@
+                `sample()`.
+        """
+        self.capacity = capacity
+        self.batch_size_B = batch_size_B
+        self.batch_length_T = batch_length_T
+
+        # The actual episode buffer.
+        self.episodes = deque()
+        # Maps (unique) episode IDs to the index under which to find this episode
+        # within our `self.episodes` deque. Note that after ejection started, the
+        # indices will NOT be changed. We will therefore need to offset these indices
+        # by the number of episodes that have already been ejected.
+        self.episode_id_to_index = {}
+        # The number of episodes that have already been ejected from the buffer
+        # due to reaching capacity. This is the offset, which we have to subtract
+        # from the episode index to get the actual location within `self.episodes`.
+        self._num_episodes_ejected = 0
+
+        # List storing all index tuples: [eps_idx, ts_in_eps_idx], where ...
+        # `eps_idx - self._num_episodes_ejected' is the index into self.episodes.
+        # `ts_in_eps_idx` is the timestep index within that episode
+        #  (0 = 1st timestep, etc..).
+        # We sample uniformly from the set of these indices in a `sample()`
+        # call.
+        self._indices = []
+
+        # The size of the buffer in timesteps.
+        self.size = 0
+        # How many timesteps have been sampled from the buffer in total?
+        self.sampled_timesteps = 0
+
+    @override(ReplayBufferInterface)
+    def __len__(self) -> int:
+        return self.get_num_timesteps()
+
+    @override(ReplayBufferInterface)
+    def add(self, batch: SampleBatchType, **kwargs) -> None:
+        """Converts the incoming SampleBatch into a number of _Episode objects.
+
+        Then adds these episodes to the internal deque.
+        """
+        if isinstance(batch, MultiAgentBatch):
+            raise ValueError(
+                "`EpisodeReplayBuffer` cannot operate on MultiAgentBatches yet! "
+                "For single-agent use only."
+            )
+
+        episode_slices = batch.split_by_episode()
+        episodes = [
+            _Episode.from_sample_batch(eps_slice)
+            for eps_slice in episode_slices
+        ]
+
+        for eps in episodes:
+            self.size += len(eps)
+            # Ongoing episode, concat to existing record.
+            if eps.id_ in self.episode_id_to_index:
+                buf_idx = self.episode_id_to_index[eps.id_]
+                existing_eps = self.episodes[buf_idx - self._num_episodes_ejected]
+                old_len = len(existing_eps)
+                self._indices.extend([(buf_idx, old_len + i) for i in range(len(eps))])
+                existing_eps.concat_episode(eps)
+            # New episode. Add to end of our episodes deque.
+            else:
+                self.episodes.append(eps)
+                buf_idx = len(self.episodes) - 1 + self._num_episodes_ejected
```

Comment:
Even though this seems to work, it seems weird to me.
On the first call, this line should look as follows:

```
buf_idx = len(self.episodes) - 1 + self._num_episodes_ejected
-1           =              0                - 1 +                      0
``
So the first buf_idx will be -1. Could it be that, even though it is working, the indexing is off somehow?

Reply:
No, the first index will be 0:

```
len(self.episodes) is already 1 at that time.
-1 -> 0
then + 0 for the `_num_episodes_evicted` (none yet).
-> 0
```

Reply:
There is an append one line above:
```
self.episodes.append(eps)
```

Reply:
Duh. Sorry.